File: python/tests/test_cross_sections.py
import pytest
import math
from conftest import SimulationAdapter

@pytest.mark.parametrize("version", ["native_version", "numpy_version", "numba_version"])
def test_phelps_cross_sections(version):
    adapter = SimulationAdapter(version)
    sim = adapter.create_state()
    cs = adapter.cs_mod

    adapter.setup_cross_sections(sim)

    idx_50 = int(50.0 / cs.DE_CS)
    total_macro = (sim.sigma[cs.E_ELA][idx_50] + sim.sigma[cs.E_EXC][idx_50] + sim.sigma[cs.E_ION][idx_50]) * cs.GAS_DENSITY

    print(f"\n[{version}] Przekroje czynne zderzen elektronow dla energii 50.0 eV:")
    print(f"  Elastyczne (sigma[E_ELA]): {sim.sigma[cs.E_ELA][idx_50]:10.2e} m^2")
    print(f"  Wzbudzenie (sigma[E_EXC]): {sim.sigma[cs.E_EXC][idx_50]:10.2e} m^2")
    print(f"  Jonizacja  (sigma[E_ION]): {sim.sigma[cs.E_ION][idx_50]:10.2e} m^2")
    print(f"  Suma makroskopowa total_e: {sim.sigma_tot_e[idx_50]:10.5f} 1/m (oczekiwana = {total_macro:10.5f} 1/m)")

    assert math.isclose(sim.sigma_tot_e[idx_50], total_macro, rel_tol=1e-7)

File: C/tests/test_cross_sections.cc
#include <gtest/gtest.h>
#include "test_helpers.h"

class CrossSectionTest : public ::testing::Test {
protected:
    void SetUp() override {
        set_electron_cross_sections_ar();
        set_ion_cross_sections_ar();
        calc_total_cross_sections();
    }
};

TEST_F(CrossSectionTest, PhelpsPetrovicArValues) {
    // Przekroje czynne dla elektronów (wszystkie wartości w 1e-20 m^2)
    // 0.1 eV -> poniżej progu na wzbudzenie i jonizację (tylko elastyczne)
    int idx_0_1 = (int)(0.1 / DE_CS);
    EXPECT_GT(sigma[E_ELA][idx_0_1], 0.0);
    EXPECT_NEAR(sigma[E_EXC][idx_0_1], 0.0, 1e-15);
    EXPECT_NEAR(sigma[E_ION][idx_0_1], 0.0, 1e-15);
    
    // 50.0 eV -> powyżej wszystkich progów
    int idx_50 = (int)(50.0 / DE_CS);
    EXPECT_GT(sigma[E_ELA][idx_50], 0.0);
    EXPECT_GT(sigma[E_EXC][idx_50], 0.0);
    EXPECT_GT(sigma[E_ION][idx_50], 0.0);
    
    // Suma przekrojów
    double total_macro = (sigma[E_ELA][idx_50] + sigma[E_EXC][idx_50] + sigma[E_ION][idx_50]) * GAS_DENSITY;
    EXPECT_NEAR(sigma_tot_e[idx_50], total_macro, 1e-12);
}